Add Semaphore extension

lr4d opened the pull request:
Supersedes #2690

Feedback is appreciated

mrocklin commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+        # We should make sure that the client is already properly registered with the scheduler
+        # otherwise the lease validation will mop up every acquired lease immediately. That's mostly relevant for tests
+        while client not in self.scheduler.clients:
+            await asyncio.sleep(0)
```
Should this be a non-zero value?

fjetter replied:
I believe this edge case is mostly triggered in test scenarios. Non-zero should be fine either way.

mrocklin commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+                self.scheduler.loop.add_callback(
+                    self._release_value, name, client, identifier
+                )
+            else:
```
Can we do this immediately?  Why do we need to do it in a future cycle?

fjetter replied:
Should be fine to convert to an ordinary function. I believe once there was a version with some waiting code and this was never converted back.

mrocklin commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+            else:
+                raise ValueError("Semaphore released too many times.")
+
+    async def _release_value(self, name, client, identifier):
```
Does this need to be `async def`?  We don't await in here, so could we make this a normal function?

mrocklin commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.leases = defaultdict(deque)
+        self.events = defaultdict(tornado.locks.Event)
```
We recently switched to using `asyncio.Event`s rather than `tornado.locks.Event`s.  Is this easy to change?

mrocklin commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+                name=self.name,
+                max_leases=max_leases,
+            )
+            self._started = True
```
Maybe this should be an awaitable like `asyncio.sleep(0)` just in case someone awaits the Semaphore later?

(people sometime mix sync and async APIs)

mrocklin commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+        self.__init__(name=name, client=client, max_leases=max_leases)
+
+    def close(self):
+        self.client.sync(self.client.scheduler.semaphore_close, name=self.name)
```
```suggestion
        return self.client.sync(self.client.scheduler.semaphore_close, name=self.name)
```

We need to return the result here.  If we're in async mode then this will be an `asyncio.Future`, and the caller will need to await it.

mrocklin commented on distributed/tests/test_semaphore.py:
```diff
@@ -0,0 +1,156 @@
+
+@gen_cluster(client=True)
+async def test_semaphore(c, s, a, b):
+    semaphore = await Semaphore(max_leases=2, name="exasol_db")
```
Why the name `exasol_db`?

fjetter replied:
That slipped in for educational purposes, I believe. We should revert this.

Background (other than we're using an Exasol)

We had a discussion about the naming of arguments and where discussing about the `name` argument and whether or not it should be named `resource` since the semaphore is likely used to limit access to a resource

lr4d replied:
I renamed the arguments `id` and `name` to `instance_id` and `resource_name` respectively

mrocklin replied:
Note that variables, locks, and queues all use a `name=` parameter.  If this parameter fulfills a similar role then it would be good to keep the name `name` for consistency.  Similarly Client uses `id` for instance id.  We might want to stick to that as well.

lr4d replied:
I'll revert that for consistency then. I still think it wouldn't hurt to refactor that sometime for easier understanding of the code

mrocklin commented on distributed/tests/test_semaphore.py:
```diff
@@ -0,0 +1,156 @@
+    await sem.release()
+
+
+def test_timeout_sync():
```
```suggestion
def test_timeout_sync(c, s, a, b):
```

I believe that there are pytest fixtures that now replace the cluster/client manual construction.

lr4d replied:
```
  def test_timeout_sync(c, s, a, b):
E       fixture 'c' not found
```

fjetter replied:
Might be that you need to import it first. With fixtures defined in a proper module, the rules are slightly different compared to `conftest` defined fixtures

lr4d replied:
ah my bad

mrocklin replied:
This is in test_client.py

```python
from distributed.utils_test import (  # noqa: F401
    client as c,
    s,
    a,
    b,
)
```

mrocklin commented on distributed/tests/test_semaphore.py:
```diff
@@ -0,0 +1,156 @@
+                    client.set_metadata("locked", True)
+                    sleep(0.01)
+                    assert client.get_metadata("locked") is True
+                    client.set_metadata("locked", False)
```
Did you want to run this function many times concurrently?

fjetter replied:
Yes, somewhere a line is lost where `f` is called using, e.g. a `client.map`

mrocklin commented on distributed/tests/test_semaphore.py:
```diff
@@ -0,0 +1,156 @@
+    ):
+        sem = await Semaphore(name="x", max_leases=2)
+        await sem.acquire()  # leases: 2 - 1 = 1
+        semY = Semaphore(name="y")
```
Did you want to await this one?

mrocklin commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+import asyncio
+import dask
+import tornado.locks
+import tornado.queues
```
This doesn't appear to be used.  My guess is that you have some excess imports here.

fjetter commented on distributed/tests/test_semaphore.py:
```diff
@@ -0,0 +1,156 @@
+
+
+def test_lock_name_only():
+    with cluster() as (scheduler, workers):
```
This could also use the async interface, couldn't it?

fjetter commented:
cc @pacman82

dhirschfeld commented:
IIUC this closes #864

lr4d commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,259 @@
+        self.__init__(name=name, client=client, max_leases=max_leases)
+
+    def close(self):
+        self.client.sync(self.client.scheduler.semaphore_close, name=self.name)
```
`scheduler.semaphore_close` is not implemented

mrocklin replied:
When we create semaphores on the client side do we create any state on the scheduler side?  Is there a need to clean up that state if the client destroys its semaphores?

If there is no need for a close method, then great, we should probably remove this entire method.  If there is some state that we would like to clean up then we should probably do something here?

lr4d replied:
I'll look into it

lr4d commented:
My pleasure, @mrocklin . I'll address the review comments in the upcoming days. Any idea why the tests for Python 3.8 fail on Windows?

lr4d commented:
I'm looking into an additional test which is currently failing and I have not yet figured out why.

wrt the Semaphore.close method, I don't see _that_ much value in implementing this at the moment, am I missing something?

mrocklin commented:
> My pleasure, @mrocklin . I'll address the review comments in the upcoming days. Any idea why the tests for Python 3.8 fail on Windows?

I'm not seeing that currently.  There are occasionally intermittently failing tests.  Note that currently Linux 3.6-3.7 are failing with TLS issues.

> wrt the Semaphore.close method, I don't see that much value in implementing this at the moment, am I missing something?

In many other situations we want to make sure that things get cleaned up.  It might be good to test that whatever collections this system maintains are cleared out if there are no active semaphores.  We just want to make sure that we don't accrue state continuously.

```python
async with Semaphore(name=...):
    ...

assert not scheduler.semaphore_state
```

lr4d commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,327 @@
+            if name in self.leases and identifier in self.leases[name]:
+                self._release_value(name, client, identifier)
+            else:
+                raise ValueError(
```
do we not care if an exception is raised here or would a warning/log message be more appropriate so as to not block further execution? cc @fjetter

mrocklin replied:
This code is triggered by a call from the Client side.  The mechanism that Dask uses to do these sorts of remote calls should serialize the error, send it to the client, and reraise it there.  This would be good to verify though.  Maybe something like the following?

```python
semaphore.acquire()
semaphore.release()

with pytest.raises(ValueError) as info:
    semaphore.release()

assert "already released" in str(info.value)
```

?

fjetter replied:
I think the question is not about how the exception is handled but rahter should we raise an exception or ignore this?

I'm inclined to say this should **not** be an exception. This path is encountered in two scenarios:
1. Releasing too often
2. Semaphore closed already

I could imagine both scenarios to happen in a distributed system due to weird timings and I wouldn't want my application to crash while cleaning up. There are different philosophies for this, I believe and I don't have a very strong opinion here.

mrocklin replied:
What is the behavior observed by the user/Client?  Is it the same as they would see with a typical Python Semaphore?  If so, then I'm happy.

lr4d replied:
Described this in the semaphore docstring under notes

lr4d commented:
> In many other situations we want to make sure that things get cleaned up. It might be good to test that whatever collections this system maintains are cleared out if there are no active semaphores. We just want to make sure that we don't accrue state continuously.

I've implemented a close method and added a line to make sure we remove all state from a client that has been "released". Would appreciate any further feedback

lr4d commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,327 @@
+
+    def _semaphore_exists(self, name):
+        if name not in self.max_leases:
+            logger.exception(f"Semaphore `{name}` not known or already closed.")
```
@mrocklin how about the handling of this error? Context: the semaphore has been closed using `Semaphore.close` and a client attempts to acquire/release/close the closed semaphore. Would the expected behavior be that there is an exception raised?

fjetter replied:
I would not log or raise here since the behaviour should be depending on where this check is performed. I think it's ok to encapsulate the _exists_ check in a method since what data structure needs to be checked is not trivial

pacman82 replied:
I'm in agreement with @fjetter on this one. The post condition is met and the semaphore released. No need to raise an exception.

mrocklin replied:
To make this concrete from the Client perspective, is this the situation that we're considering?

```python
# client 1
semaphore = Semaphore(name="my-resource")
# do things with semaphore
semaphore.close()
```

```python
# client 2
semaphore = Semaphore(name="my-resource")
ValueError("Semaphore already closed")
```

If so I agree that this shouldn't occur.  Perhaps this isn't what we're discussing though.  If folks here are able to spell out what the result of this code is that would help in review (sorry, there is a large backlog to go through, so I'm rushing through reviews today)

lr4d replied:
changed this behavior in https://github.com/dask/distributed/pull/3573/commits/7a519279a2539203f5dd2f18c0506365b6f28e4e

@mrocklin no, it is for the case when any client closes a semaphore and then any client tries to acquire, release or close that semaphore. See the commit I linked for the behavior for each of these operations now. 

Happy to edit that commit one final time if we can get to a consensus

fjetter commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,327 @@
+    ):
+        with log_errors():
+            if not self._semaphore_exists(name):
+                return
```
1. Acquire should return a boolean. The boolean is well defined and it is clear what's going on.
2. I would argue that an acquire for a not yet create semaphore should raise

lr4d replied:
Would you also argue that it should raise for a semaphore that has already been closed?

mrocklin replied:
Without knowing too much about the situation here, I would default to whatever the behavior is for the normal Python Semaphore

fjetter replied:
1. The python semaphore always returns bool
2. The "does not exist" does not exist in the python standard lib (since it's not distributed) so there are no references. I would argue this should raise in that scenario. 

this scenario would happen if

```
# client 1
semaphore = Semaphore(name="my-resource")

# client 2
semaphore = Semaphore(name="my-resource")

# client 1
semaphore.close()

# client 2
semaphore.acquire()
ValueError("Semaphore already closed")
```

as pointed out in the above [thread](https://github.com/dask/distributed/pull/3573/files/17ba039ff95dc84cb054e1d2872172bc1c3950ca#r396113995)

I believe this scenario should properly raise and we should make sure that Semaphore close should only be called by the process which is responsible for scheduling, e.g.

```
sem = Semaphore()
graph = my_dask_comp(sem=sem)
graph.compute()
sem.close()
```

dhirschfeld commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,327 @@
+    Examples
+    --------
+    >>> sem = Semaphore(max_leases=2, name='my_database')  # doctest: +SKIP
+    >>> sem.acquire(timeout=0.1)  # doctest: +SKIP
```
Perhaps the example should advertise the context-manager api as that's generally preferred to manually cleaning up resources.

dhirschfeld commented:
I'm curious if this works with [`fire_and_forget`](https://docs.dask.org/en/latest/futures.html#fire-and-forget)

In my usage a client will connect, fire off 100 tasks then exit. `fire_and_forget` ensures that even though the client goes away all the submitted tasks will be run. What I'd like is to be able to use the `Semaphore` to also limit the number of tasks running in parallel.

Just wondering if `Semaphore` implements some black magic that makes that possible?

What I do currently is to have the client submit a "manager" task which submits the 100 tasks itself using a `worker_client` and the `as_completed` trick. I imagine that I'd still have to use a similar indirection with `Semaphore`?

pacman82 commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,327 @@
+        # We should make sure that the client is already properly registered with the scheduler
+        # otherwise the lease validation will mop up every acquired lease immediately. That's mostly relevant for tests
+        while client not in self.scheduler.clients:
+            await asyncio.sleep(0.0005)  # This value is set somewhat arbitrarily
```
What if the client get's deleted during waiting for a pending semaphore (can this happen?). In such a case it might block for the complete duration of a timeout in context of `acquire`.

lr4d replied:
confirmed and removed

mrocklin commented:
> What I'd like is to be able to use the Semaphore to also limit the number of tasks running in parallel.

It won't limit the number of tasks, but it will cause them to block.  Semaphores and `fire_and_forget` don't conflict with each other in any way.  So after this PR you should be able to launch all of your tasks at once, but have all of those tasks wait on the Semaphore.   That should give you the kind of rate limiting that you're looking for without doing any work on the client side.

dhirschfeld commented:
Thanks, I look forward to trying it out!

lr4d commented:
@fjetter Seems reasonable to me. I'm curious if this change could interfere with somebody else's usual method of execution (assuming they have valid reasons for doing so). Would also like to hear others' feedback on this

fjetter commented:
@mrocklin from my POV this is ready to merge. I believe there are a few more things to clean up (e.g. the above changes in interface) but I don't think that should be a blocker since the current implementation is ready for users to experiment with.

There are a few follow up tasks for which I would create github issues to track
* Semaphore close as contextmanager
* Client request timeouts separate to lease timeout
* Semaphore observability using prometheus metrics
* Lock uses semaphore implementation as backend

Is there a best practice of flagging new features as experimental to let users know that 
A) it might still be buggy, don't expect too much 
B) there may be subtle breaking changes on the horizon

fjetter commented on distributed/semaphore.py:
```diff
@@ -0,0 +1,336 @@
+    >>> # the context manager
+    >>> with sem:
+    >>>     pass  # use the protected resource
+    >>> sem.close()  # Close resource (clearing any residual state)
```
```suggestion
    >>> def access_ressource(s, sem):
    >>>     # This automatically acquires a lease from the semaphore (if available) which will be
    >>>     # released when leaving the context manager.
    >>>     with sem:
    >>>         pass
    >>>    
    >>> futures = client.map(access_ressource, range(10), sem=sem)
    >>> client.gather(futures)
    >>> # Once done, close the semaphore to clean up the state on scheduler side.
    >>> sem.close()
    
```

This example shows the (recommended) usage where the semaphore is initialised on client side and not in the worker. This usage makes the cleanup more straight forward

mrocklin commented:
> @mrocklin from my POV this is ready to merge

In that case I'm happy to defer to your judgement here.  I've been quite busy recently and so I haven't had time to give this PR the full attention that it deserves.  Merge at will.

Also, when merging, I recommend the following practice:

-  We use "Squash and merge" for PRs, unless the commit history is particularly concise and valuable (it has been many many months since I did anything other than squash and merge)
-  Similarly, unless the commit history is particularly informative, I recommend removing the list summary that is automatically generated.  Instead, I often copy in the header comment (if it's informative) or write up something short myself

> Is there a best practice of flagging new features as experimental to let users know that
> A) it might still be buggy, don't expect too much
> B) there may be subtle breaking changes on the horizon

The simplest thing is probably to put a note in the docstring.  We don't currently have any formal warning mechanism for this.  

You might also consider a small documentation PR to dask/dask, which holds the user documentation today.  There is a section on coordination primitives that includes Locks and Queues and such where this would fit in well.
